analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs

```diff
@@ -433,6 +433,48 @@ public void Branching_LearnsObjectConstraint_TypePattern_Negated(string expressi
                 .And.ContainSingle(x => x != null && x.HasConstraint(ObjectConstraint.NotNull));
         }
 
+        [DataTestMethod]
+        [DataRow("arg is object o")]
+        [DataRow("arg is string s")]
+        [DataRow("arg is not not object o")]
+        [DataRow("arg is not not string s")]
+        [DataRow("!(arg is not object o)")]
+        public void Branching_LearnsObjectConstraint_DeclarationPattern(string expression)
+        {
+            var validator = CreateIfElseEndValidatorCS(expression, OperationKind.DeclarationPattern);
+            validator.ValidateTag("If", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("Else", x => x.Should().BeNull("it could be null or any other type"));
+            validator.TagValues("End").Should().HaveCount(2)
+                .And.ContainSingle(x => x == null)
+                .And.ContainSingle(x => x != null && x.HasConstraint(ObjectConstraint.NotNull));
+        }
+
+        [DataTestMethod]
+        [DataRow("arg is var o")]
+        [DataRow("arg is int i")]
```

Comment:
This is wrong. We learn that arg is not null in "If".

Reply:
We don't build object constraints for value types

Reply:
Once we rewrite the CBDE rule, we'll start building int constraints about their range. Until then, this should be fine

Reply:
> We don't build object constraints for value types

We would put the "NotNull" constraint on "arg" which isn't a value type.
I think for consistency we should support boxing/unboxing now because in recursive patterns we already do so. It also seems simpler to add support than not. Let's wait for the recursive pattern PR to get merged, so we have a better understanding of what needs to be changed here and for constant patterns.

Reply:
Done